File: modules/virtual-people-core-serving/src/main/cc/wfa/virtual_people/core/model/model_node.cc
// Copyright 2021 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "wfa/virtual_people/core/model/model_node.h"

#include <memory>

#include "absl/status/statusor.h"
#include "wfa/virtual_people/common/model.pb.h"
#include "wfa/virtual_people/core/model/branch_node_impl.h"
#include "wfa/virtual_people/core/model/population_node_impl.h"
#include "wfa/virtual_people/core/model/ranked_population_node_impl.h"
#include "wfa/virtual_people/core/model/stop_node_impl.h"

namespace wfa_virtual_people {

absl::StatusOr<std::unique_ptr<ModelNode>> ModelNode::Build(
    const CompiledNode& config,
    absl::flat_hash_map<uint32_t, std::unique_ptr<ModelNode>>& node_refs) {
  if (config.type_case() == CompiledNode::TypeCase::kBranchNode) {
    return BranchNodeImpl::Build(config, node_refs);
  }
  return ModelNode::Build(config);
}

absl::StatusOr<std::unique_ptr<ModelNode>> ModelNode::Build(
    const CompiledNode& config) {
  switch (config.type_case()) {
    case CompiledNode::TypeCase::kBranchNode: {
      absl::flat_hash_map<uint32_t, std::unique_ptr<ModelNode>> node_refs;
      return BranchNodeImpl::Build(config, node_refs);
    }
    case CompiledNode::TypeCase::kStopNode:
      return StopNodeImpl::Build(config);
    case CompiledNode::TypeCase::kPopulationNode:
      return PopulationNodeImpl::Build(config);
    case CompiledNode::TypeCase::kRankedPopulationNode:
      return RankedPopulationNodeImpl::Build(config);
    default:
      return absl::InvalidArgumentError("Node type is not set.");
  }
}

ModelNode::ModelNode(const CompiledNode& node_config)
    : name_(node_config.name()),
      from_model_builder_config_(
          node_config.debug_info().directly_from_model_builder_config()) {}

}  // namespace wfa_virtual_people

File: modules/virtual-people-core-serving/src/main/kotlin/org/wfanet/virtualpeople/core/model/ModelNode.kt
// Copyright 2022 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.wfanet.virtualpeople.core.model

import org.wfanet.virtualpeople.common.CompiledNode
import org.wfanet.virtualpeople.common.CompiledNode.TypeCase
import org.wfanet.virtualpeople.common.LabelerEvent

/**
 * The Kotlin implementation of [CompiledNode] proto. Each node in the model tree will be converted
 * to a [ModelNode]. Except for debugging purposes, this should be used by VID Labeler only.
 *
 * This is a base class for all model node classes. Never add any behavior here. Only fields
 * required for all model node classes should be added here.
 */
sealed class ModelNode(nodeConfig: CompiledNode) {
  private val name: String
  private val fromModelBuilderConfig: Boolean

  init {
    name = nodeConfig.name
    fromModelBuilderConfig = nodeConfig.debugInfo.directlyFromModelBuilderConfig
  }

  /** Applies the node to the [event]. */
  abstract fun apply(event: LabelerEvent.Builder)

  companion object {
    /**
     * Always use [ModelNode]::Build to get a [ModelNode] object. Users should never call the
     * factory function or constructor of the derived class directly.
     *
     * @param [nodeRefs] the mapping from indexes to the [ModelNode] objects, which should contain
     *   the child nodes referenced by indexes. Throws an error if any child node referenced by
     *   index is not found in [nodeRefs].
     */
    fun build(config: CompiledNode, nodeRefs: MutableMap<Int, ModelNode>): ModelNode {
      return if (config.typeCase == TypeCase.BRANCH_NODE) {
        BranchNodeImpl.build(config, nodeRefs)
      } else {
        build(config)
      }
    }

    /** Build nodes with no index references in the sub-tree. */
    fun build(config: CompiledNode): ModelNode {
      @Suppress("WHEN_ENUM_CAN_BE_NULL_IN_JAVA") // Proto enum fields are never null.
      return when (config.typeCase) {
        TypeCase.BRANCH_NODE -> BranchNodeImpl.build(config, mutableMapOf())
        TypeCase.STOP_NODE -> StopNodeImpl.build(config)
        TypeCase.POPULATION_NODE -> PopulationNodeImpl.build(config)
        TypeCase.RANKED_POPULATION_NODE -> RankedPopulationNodeImpl.build(config)
        TypeCase.TYPE_NOT_SET -> error("Node type is not set.")
      }
    }
  }
}
